tls/s2n_client_hello.c

```diff
@@ -232,6 +236,74 @@ int s2n_client_hello_recv(struct s2n_connection *conn)
     return 0;
 }
 
+static int s2n_parsed_extensions_compare(const void *p, const void *q)
+{
+
+    const struct s2n_client_hello_parsed_extension *left = (const struct s2n_client_hello_parsed_extension *) p;
+    const struct s2n_client_hello_parsed_extension *right = (const struct s2n_client_hello_parsed_extension *) q;
+
+    return left->extension_type - right->extension_type;
+}
+
+static int s2n_populate_client_hello_extensions(struct s2n_client_hello *ch)
+{
+    if (ch->parsed_extensions == NULL) {
+        notnull_check(ch->parsed_extensions = s2n_array_new(sizeof(struct s2n_client_hello_parsed_extension)));
+    }
+
+    struct s2n_stuffer in;
+
+    GUARD(s2n_stuffer_init(&in, &ch->extensions));
+    GUARD(s2n_stuffer_write(&in, &ch->extensions));
+
+    while (s2n_stuffer_data_available(&in)) {
+        uint16_t ext_size, ext_type;
+
+        GUARD(s2n_stuffer_read_uint16(&in, &ext_type));
+        GUARD(s2n_stuffer_read_uint16(&in, &ext_size));
+
+        struct s2n_client_hello_parsed_extension *parsed_extension = s2n_array_add(ch->parsed_extensions);
```

Comment:
should we check to ensure there isn't already an extension of type ext_type before adding it here?

Seeing that you build the array and then sort, perhaps you could just do it after you sort it and fail the handshake if we have duplicates.

that said, this is could be a follow up PR to close out https://github.com/awslabs/s2n/issues/728

Reply:
Yup that is the idea, I will do it as a follow up PR along with few other minor changes.